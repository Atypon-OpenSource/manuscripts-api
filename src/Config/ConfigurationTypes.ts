/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { RSA_JWK } from 'pem-jwk'

import { ClientApplication } from '../Models/ClientApplicationModels'

export enum Environment {
  Test = 'test',
  Development = 'development',
  Production = 'production',
}

/**
 * Represents an API configuration.
 */
export type APIConfiguration = {
  readonly port: number
  readonly hostname: string
}

/**
 * Represents a database configuration.
 */
export enum BucketKey {
  User = 'user',
  Project = 'project',
}

export type DatabaseConfiguration = {
  readonly buckets: { [name in BucketKey]: string }
}

export type ScopedAccessTokenConfiguration = {
  readonly name: string
  readonly secret: string
  readonly publicKeyPEM: string | null
  readonly publicKeyJWK: RSA_JWK | null
  readonly expiry: number
  readonly identifier: string // maps to the "kid" property when represented in JWK.
}

/**
 * Represents authentication configuration.
 */
export type AuthConfiguration = {
  readonly jwtSecret: string
  readonly skipVerification: boolean
  readonly hashSaltRounds: number
  readonly serverSecret: string
  readonly enableNonConnectAuth: boolean
}

/**
 * Represents email sending configuration.
 */
export type EmailConfiguration = {
  readonly fromAddress: string
  readonly fromBaseURL: string
}

export type ClientApplicationsConfiguration = {
  readonly knownClientApplications: Array<ClientApplication>
}

export type ServerConfiguration = {
  readonly allowedCORSOrigins: ReadonlyArray<string>
}

export interface EnvironmentLike {
  [key: string]: string | undefined
}

export interface PressroomConfiguration {
  readonly baseurl: string
  readonly apiKey: string
}

export interface ConfigurationContainer {
  readonly API: APIConfiguration
  readonly DB: DatabaseConfiguration
  readonly auth: AuthConfiguration
  readonly email: EmailConfiguration
  readonly server: ServerConfiguration
  readonly apps: ClientApplicationsConfiguration
  readonly scopes: ScopedAccessTokenConfiguration[]
  readonly pressroom: PressroomConfiguration
}
