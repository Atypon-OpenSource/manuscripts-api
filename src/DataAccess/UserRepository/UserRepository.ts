/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Prisma } from '@prisma/client'

import { DatabaseError } from '../../Errors'
import { INewUser, IUpdateUser, User, userForRow, UserRow } from '../../Models/UserModels'
import { IUserRepository } from '../Interfaces/IUserRepository'
import { UserQueryCriteria } from '../Interfaces/QueryCriteria'
import { SQLRepository } from '../SQLRepository'
import { maxLength, required, validEmail } from '../validators'

/**
 * Manages user persistent storage operations.
 */
export class UserRepository
  extends SQLRepository<User, INewUser, Partial<IUpdateUser>, UserQueryCriteria>
  implements IUserRepository
{
  /**
   * Returns document type
   */
  public get documentType(): string {
    return 'User'
  }

  public buildSchemaDefinition(): any {
    return {
      _id: {
        type: 'string',
        auto: 'uuid',
        readonly: true,
      },
      name: {
        type: 'string',
        validator: (val: string) => {
          required(val, 'name')
          maxLength(val, 100, 'name')
        },
      },
      email: {
        type: 'string',
        validator: (val: string) => {
          required(val, 'email')
          validEmail(val, 'email')
          maxLength(val, 100, 'email')
        },
      },
      connectUserID: {
        type: 'string',
      },
      deleteAt: {
        type: 'number',
      },
    }
  }

  /**
   * Builds a user model from a user row object.
   */
  public buildModel(user: UserRow): User {
    return userForRow(user)
  }

  /**
   * Returns users based on the value of property `deleteAt`.
   */
  public getUsersToDelete(): Promise<User[] | null> {
    const Q = {
      data: {
        path: ['deleteAt'],
        lte: new Date(),
      },
    }
    return new Promise((resolve, reject) => {
      return this.database.bucket
        .query(Q)
        .catch((error: Prisma.PrismaClientKnownRequestError) =>
          reject(
            DatabaseError.fromPrismaError(
              error,
              `Error getting users to delete ${this.documentType}`,
              JSON.stringify(Q)
            )
          )
        )
        .then((docs: any) => {
          if (docs && docs.length) {
            const users = docs.map((user: any) => {
              return this.buildModel(user)
            })
            return resolve(users)
          } else {
            return resolve(null)
          }
        })
    })
  }
}
