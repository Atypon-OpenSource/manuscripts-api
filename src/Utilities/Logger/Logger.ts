/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { WinstonLogger } from './Winston'

const sharedLogger = new WinstonLogger()

class Logger {
  /**
   * Logs information data.
   */
  public info(msg: string, meta?: any): void {
    sharedLogger.info(msg, meta)
  }

  /**
   * Logs errors.
   */
  public error(msg: string, meta?: any): void {
    sharedLogger.error(msg, meta)
  }

  /**
   * Logs debug data
   */
  public debug(msg: string, meta?: any): void {
    sharedLogger.debug(msg, meta)
  }
}

export const log = new Logger()
