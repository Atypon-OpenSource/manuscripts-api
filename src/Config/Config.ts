/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as path from 'path'
import { load } from 'dotenv-safe'
import { isString, isNumber } from '../util'
import { CreateBucketOptions } from 'couchbase'
import {
  clientApplicationsFromSplitString,
  scopeConfigurationsFromSplitString,
} from '../Models/ClientApplicationModels'
import { existsSync, readFileSync } from 'fs'
import { log } from '../Utilities/Logger'
import { ConfigurationError } from '../Errors'
import {
  APIConfiguration,
  DatabaseConfiguration,
  AuthConfiguration,
  AWSConfiguration,
  GoogleConfiguration,
  EmailConfiguration,
  GatewayConfiguration,
  ServerConfiguration,
  ClientApplicationsConfiguration,
  EnvironmentLike,
  Environment,
  ConfigurationContainer,
  IAMConfiguration,
  ScopedAccessTokenConfiguration,
  LiteratumConfiguration,
  PressroomConfiguration,
  TemplateConfiguration,
  ShacklesConfiguration,
} from './ConfigurationTypes'
import { normalizeURL } from './normalize-url'

const isUTF8 = require('is-utf8')

/**
 * Configuration string validator:
 * passes input parameter through as return value
 * if it is a non-empty string, throws otherwise.
 */
function getString(value: any, key: string): string {
  if (!isString(value) || value.length === 0) {
    throw new ConfigurationError(key, value)
  }
  return value
}

function getArray(value: string, key: string, separator: string): string[] {
  if (value === '') {
    return []
  }
  const arrayParts = value.split(separator)
  for (const part of arrayParts) {
    if (!isString(part) || part.length === 0) {
      throw new ConfigurationError(key, value)
    }
  }
  return arrayParts
}

function getOptionalArray(envVar: string | undefined, key: string, separator: string): string[] {
  if (envVar === undefined || envVar === '') {
    return []
  }
  const value = getString(envVar, key)
  const arrayParts = value.split(separator)
  for (const part of arrayParts) {
    if (!isString(part) || part.length === 0) {
      throw new ConfigurationError(key, value)
    }
  }
  return arrayParts
}

// undefined intentionally here,
// in order to avoid creating an object with null values
// (which the relevant 3rd party code may not interpret similarly).
function getStringOptional(value: any): string | undefined {
  if (!isString(value) || value.length === 0) {
    return undefined
  }
  return value
}

/**
 * Configuration number validator:
 * passes input parameter through as return value
 * if it is a number, return null otherwise.
 *
 * The use of undefined (instead of null) is intentional, in order to avoid
 * passing objects with null values to 3rd party code (which may not interpret them correctly).
 */
function getNumberOptional(value: any): number | undefined {
  const numberVal = Number(value)
  if (!isNumber(numberVal) || isNaN(numberVal) || !isFinite(numberVal)) {
    return undefined
  }
  return numberVal
}

function getNumber(value: any, key: string, allowMissing?: boolean): number {
  const numberVal = Number(value)
  if (!allowMissing && (!isNumber(numberVal) || isNaN(numberVal) || !isFinite(numberVal))) {
    throw new ConfigurationError(key, value)
  }
  return numberVal
}

export class Configuration implements ConfigurationContainer {
  readonly API: APIConfiguration
  readonly DB: DatabaseConfiguration
  readonly auth: AuthConfiguration
  readonly AWS: AWSConfiguration
  readonly google: GoogleConfiguration
  readonly IAM: IAMConfiguration
  readonly email: EmailConfiguration
  readonly gateway: GatewayConfiguration
  readonly server: ServerConfiguration
  readonly apps: ClientApplicationsConfiguration
  readonly scopes: ScopedAccessTokenConfiguration[]
  readonly literatum: LiteratumConfiguration
  readonly pressroom: PressroomConfiguration
  readonly shackles: ShacklesConfiguration
  readonly template: TemplateConfiguration

  constructor(env: EnvironmentLike) {
    const bucketOptions = this.createBucketOptions(env)

    this.API = {
      port: Number(env.APP_PORT),
      oauthStateEncryptionKey: getString(
        env.APP_OAUTH_STATE_ENCRYPTION_KEY,
        'APP_OAUTH_STATE_ENCRYPTION_KEY'
      ),
      hostname: getString(env.APP_HOSTNAME_PUBLIC, 'APP_HOSTNAME_PUBLIC'),
    }

    /* istanbul ignore next */
    const skipVerification =
      env.NODE_ENV === 'development' &&
      Boolean(getNumber(env.APP_SKIP_ACCOUNT_VERIFICATION, 'APP_SKIP_ACCOUNT_VERIFICATION'))

    this.auth = {
      jwtSecret: getString(env.APP_JWT_SECRET, 'APP_JWT_SECRET'),
      skipVerification: skipVerification,
      hashSaltRounds:
        getNumber(env.APP_HASH_SALT_ROUNDS, 'APP_HASH_SALT_ROUNDS', true) ||
        (env.NODE_ENV === Environment.Production ? 10 : 3),
      serverSecret: getString(env.APP_SERVER_SECRET, 'APP_SERVER_SECRET'),
      enableNonConnectAuth: Boolean(
        getNumber(env.APP_ENABLE_NON_CONNECT_AUTH, 'APP_ENABLE_NON_CONNECT_AUTH')
      ),
    }

    const buckets = {
      user: getString(env.APP_USER_BUCKET, 'APP_USER_BUCKET'),
      data: getString(env.APP_DATA_BUCKET, 'APP_DATA_BUCKET'),
      project: getString(env.APP_DATA_BUCKET, 'APP_DATA_BUCKET'),
      state: getString(env.APP_STATE_BUCKET, 'APP_STATE_BUCKET'),
      derivedData: getString(env.APP_DERIVED_DATA_BUCKET, 'APP_DERIVED_DATA_BUCKET'),
    }

    // database initialization should not be happening without INITIALIZE_DATABASE='true',
    // except for when running tests when it should occur also when INITIALIZE_DATABASE was omitted.
    /* istanbul ignore next */
    const initialize =
      env.INITIALIZE_DATABASE === 'true' ||
      (typeof env.INITIALIZE_DATABASE === 'undefined' && env.NODE_ENV === Environment.Test)

    this.DB = {
      bucketOptions,
      buckets,
      initializeContents: initialize,
      uri: normalizeURL(getString(env.APP_DB_URI, 'APP_DB_URI')),
      username: getString(env.APP_COUCHBASE_ADMIN_USER, 'APP_COUCHBASE_ADMIN_USER'),
      password: getString(env.APP_COUCHBASE_ADMIN_PASS, 'APP_COUCHBASE_ADMIN_PASS'),
      bucketAdminPassword: getString(
        env.APP_COUCHBASE_RBAC_PASSWORD,
        'APP_COUCHBASE_RBAC_PASSWORD'
      ),
    }

    this.AWS = {
      accessKeyId: getString(env.APP_AWS_ACCESS_KEY_ID, 'APP_AWS_ACCESS_KEY_ID'),
      secretAccessKey: getString(env.APP_AWS_SECRET_ACCESS_KEY, 'APP_AWS_SECRET_ACCESS_KEY'),
      region: getString(env.APP_AWS_REGION, 'APP_AWS_REGION'),
    }

    this.google = {
      clientID: getString(env.APP_GOOGLE_CLIENT_ID, 'APP_GOOGLE_CLIENT_ID'),
      clientSecret: getString(env.APP_GOOGLE_CLIENT_SECRET, 'APP_GOOGLE_CLIENT_SECRET'),
      authCallback: normalizeURL(
        getString(env.APP_GOOGLE_AUTH_CALLBACK, 'APP_GOOGLE_AUTH_CALLBACK')
      ),
    }

    this.IAM = {
      clientID: getString(env.APP_IAM_CLIENT_ID, 'APP_IAM_CLIENT_ID'),
      authServerURL: getString(env.APP_IAM_SERVER_URL, 'APP_IAM_SERVER_URL'),
      authServerPermittedURLs: Array.from(
        new Set(
          getArray(
            getString(env.APP_IAM_PERMITTED_SERVER_URLS, 'APP_IAM_PERMITTED_SERVER_URLS'),
            'APP_IAM_PERMITTED_SERVER_URLS',
            ';'
          )
        )
      ),
      authCallbackPath: normalizeURL(
        getString(env.APP_IAM_AUTH_CALLBACK_PATH, 'APP_IAM_AUTH_CALLBACK_PATH')
      ),
      libraryURL: normalizeURL(getString(env.APP_IAM_LIBRARY_URL, 'APP_IAM_LIBRARY_URL')),
      apiServerURL: Array.from(
        new Set(
          getArray(
            getString(env.APP_API_SERVER_URL, 'APP_API_SERVER_URL'),
            'APP_API_SERVER_URL',
            ';'
          )
        )
      ),
    }

    this.email = {
      fromAddress: getString(env.APP_FROM_EMAIL, 'APP_FROM_EMAIL'),
      fromBaseURL: normalizeURL(getString(env.APP_BASE_URL, 'APP_BASE_URL')),
    }

    this.gateway = {
      cookieDomain: getString(env.APP_GATEWAY_COOKIE_DOMAIN, 'APP_GATEWAY_COOKIE_DOMAIN'),
      hostname: getString(env.APP_GATEWAY_HOSTNAME, 'APP_GATEWAY_HOSTNAME'),
      ports: {
        admin: getString(env.APP_GATEWAY_ADMIN_PORT, 'APP_GATEWAY_ADMIN_PORT'),
        public: getString(env.APP_GATEWAY_PUBLIC_PORT, 'APP_GATEWAY_PUBLIC_PORT'),
      },
    }

    const host = this.email.fromBaseURL.replace(/https{0,1}\:\/\//, '')
    const additionalOrigins = [`http://${host}`, `https://${host}`]
    this.server = {
      storeOnlySSLTransmittedCookies: Boolean(
        getNumber(
          env.APP_STORE_ONLY_SSL_TRANSMITTED_COOKIES,
          'APP_STORE_ONLY_SSL_TRANSMITTED_COOKIES'
        )
      ),
      allowedCORSOrigins: Array.from(
        new Set(
          getArray(
            getString(env.APP_ALLOWED_CORS_ORIGINS, 'APP_ALLOWED_CORS_ORIGINS'),
            'APP_ALLOWED_CORS_ORIGINS',
            ';'
          ).concat(additionalOrigins)
        )
      ), // get unique values from potentially duplicated ones.
    }

    this.literatum = {
      allowedIPAddresses: Array.from(
        new Set(
          getArray(
            getString(env.APP_ALLOWED_IP_ADDRESSES, 'APP_ALLOWED_IP_ADDRESSES'),
            'APP_ALLOWED_IP_ADDRESSES',
            ';'
          )
        )
      ),
    }

    this.apps = {
      knownClientApplications: clientApplicationsFromSplitString(
        getString(env.APP_CLIENT_APPLICATIONS, 'APP_CLIENT_APPLICATIONS'),
        ';',
        ','
      ),
    }

    this.scopes = scopeConfigurationsFromSplitString(
      getString(env.APP_CONTAINER_TOKEN_SCOPES, 'APP_CONTAINER_TOKEN_SCOPES'),
      ';',
      ','
    )

    this.pressroom = {
      baseurl: getString(env.APP_PRESSROOM_BASE_URL, 'APP_PRESSROOM_BASE_URL'),
      apiKey: getString(env.APP_PRESSROOM_APIKEY, 'APP_PRESSROOM_APIKEY'),
    }

    this.shackles = {
      baseUrl: getString(env.SHACKLES_URL, 'SHACKLES_URL'),
    }

    this.template = {
      allowedOwners: getOptionalArray(
        env.APP_COUCHBASE_ALLOWED_OWNERS,
        'APP_COUCHBASE_ALLOWED_OWNERS',
        ';'
      ),
      allowedProjects: getOptionalArray(
        env.APP_COUCHBASE_ALLOWED_PROJECTS,
        'APP_COUCHBASE_ALLOWED_PROJECTS',
        ';'
      ),
    }
  }

  static fromEnv(envExamplePath: string) {
    if (existsSync('.env') && isUTF8(readFileSync('.env'))) {
      const dotenvResult = load({
        allowEmptyValues: false,
        sample: envExamplePath,
      })

      // hack to ignore the case of missing .env
      if (dotenvResult && dotenvResult.error) {
        throw dotenvResult.error
      }
    } else {
      log.error('.env is missing or is not a UTF8 encoded text file.')
    }

    return process.env // load() call above loads .env file into process.env
  }

  public createBucketOptions(env: NodeJS.ProcessEnv): CreateBucketOptions {
    const authType = getStringOptional(env.APP_DB_AUTH_TYPE)
    const bucketType = getStringOptional(env.APP_DB_BUCKET_TYPE)
    const ramQuotaMB = getNumberOptional(env.APP_DB_RAM_QUOTA_MB)
    const replicaNumber = getNumberOptional(env.APP_DB_REPLICA_NUMBER)

    const bucketOptions: CreateBucketOptions = {}
    if (authType) {
      bucketOptions.authType = authType
    }
    if (bucketType) {
      bucketOptions.bucketType = bucketType
    }
    if (ramQuotaMB) {
      bucketOptions.ramQuotaMB = ramQuotaMB
    }
    if (replicaNumber) {
      bucketOptions.replicaNumber = replicaNumber
    }

    return bucketOptions
  }
}

const env = Configuration.fromEnv(path.join(__dirname, `../../.env`))
export const config = new Configuration(env)
