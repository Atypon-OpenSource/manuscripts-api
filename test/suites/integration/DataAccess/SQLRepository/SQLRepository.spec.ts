/*!
 * © 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Chance } from 'chance'

import { UserRepository } from '../../../../../src/DataAccess/UserRepository/UserRepository'
import { drop, seed, testDatabase, dropBucket } from '../../../../utilities/db'
import {
  validUser1,
  validUser2,
  validNewUser,
  NewUserNoId,
  validNewUser2,
} from '../../../../data/fixtures/UserRepository'
import { ValidationError, DatabaseError, NoBucketError } from '../../../../../src/Errors'
import { TEST_TIMEOUT } from '../../../../utilities/testSetup'
import { log } from '../../../../../src/Utilities/Logger'
import { setTimeout } from 'timers'
import { BucketKey } from '../../../../../src/Config/ConfigurationTypes'

jest.setTimeout(TEST_TIMEOUT)

let db: any = null
beforeAll(async () => (db = await testDatabase()))
afterAll(() => db.bucket.disconnect())

const chance = new Chance()
describe('SQLRepository', () => {
  test('documentType should match expectation', () => {
    log.info('Testing documentType…')
    const repository = new UserRepository(db)
    expect(repository.documentType).toBe('User')
  })
})

describe('SQLRepository Create', () => {
  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Project)
    await seed({ users: true })
  })

  test('should create user successfully', async () => {
    const repository = new UserRepository(db)
    const user = await repository.create(validNewUser)

    expect(user).toEqual(user)
  })

  test('should fail if error occurred - same key exists', () => {
    const chance = new Chance()
    const repository = new UserRepository(db)
    const newUser = {
      _id: 'valid-user@manuscriptsapp.com',
      email: chance.email(),
      name: chance.name(),
    }
    return expect(repository.create(newUser)).rejects.toThrowError(DatabaseError)
  })

  test.skip('should delete the created user after the expiry time pass', async () => {
    const repository = new UserRepository(db)
    const user: any = await repository.create(validNewUser2)

    expect(user).toEqual({
      _id: validNewUser2._id,
      email: validNewUser2.email,
      name: validNewUser2.name,
    })

    await new Promise((resolve, reject) =>
      setTimeout(() => {
        repository
          .getById(validNewUser2._id)
          .then((result) => {
            expect(result).toBeNull()
            resolve()
          })
          .catch((error) => reject(error))
      }, 2000)
    )
  })
})

describe('SQLRepository update', () => {
  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Project)
    await seed({ users: true })
  })

  test('should fail if the id is not specified', () => {
    const repository = new UserRepository(db)
    return expect(repository.update(NewUserNoId as any)).rejects.toThrowError(ValidationError)
  })

  test('should fail if id does not exists in the database', () => {
    const repository = new UserRepository(db)

    return expect(repository.update(validNewUser)).rejects.toThrowError(DatabaseError)
  })

  test('should fail if email is in a wrong format', () => {
    const repository = new UserRepository(db)
    const userUpdatedData = {
      _id: validUser1._id,
      name: 'New Username',
      email: 'new-email',
    }

    return expect(repository.update(userUpdatedData)).rejects.toThrowError(DatabaseError)
  })

  test('should fail if name is in longer than 100 character', () => {
    const repository = new UserRepository(db)
    const userUpdatedData = {
      _id: validUser1.email,
      name: 'New Username that is too looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong, more than 100 character, ',
      email: 'new-email@manuscriptsapp.com',
    }

    return expect(repository.update(userUpdatedData)).rejects.toThrowError(DatabaseError)
  })

  test('should fail if the document _type defined are not matched to the repository type', () => {
    const repository = new UserRepository(db)
    const userUpdatedData = {
      _type: 'Not User',
      _id: validUser1._id,
      name: 'New Username',
      email: 'new-email@manuscriptsapp.com',
    }

    return expect(repository.update(userUpdatedData)).rejects.toThrowError(ValidationError)
  })

  test('should update user successfully if the document _type is defined', async () => {
    const repository = new UserRepository(db)
    const beforeUpdate: any = await repository.getById(validUser1._id)

    expect(beforeUpdate.email).toBe(validUser1.email)
    expect(beforeUpdate.name).toBe(validUser1.name)

    const userUpdatedData = {
      _type: 'User',
      _id: 'User|valid-user@manuscriptsapp.com',
      name: 'New Username',
      email: 'new-email@manuscriptsapp.com',
    }

    const afterUpdate: any = await repository.update(userUpdatedData)

    expect(afterUpdate.email).toBe('new-email@manuscriptsapp.com')
    expect(afterUpdate.name).toBe('New Username')
  })

  test('should update user successfully if the document _type is undefined', async () => {
    const repository = new UserRepository(db)
    const beforeUpdate: any = await repository.getById(validUser1._id)

    expect(beforeUpdate.email).toBe(validUser1.email)
    expect(beforeUpdate.name).toBe(validUser1.name)

    const userUpdatedData = {
      _id: 'User|valid-user@manuscriptsapp.com',
      name: 'New Username',
      email: 'new-email@manuscriptsapp.com',
    }

    const afterUpdate: any = await repository.update(userUpdatedData)

    expect(afterUpdate.email).toBe('new-email@manuscriptsapp.com')
    expect(afterUpdate.name).toBe('New Username')
  })

  test.skip('should delete the updated user after the expiry time pass', async () => {
    const repository = new UserRepository(db)

    const userUpdatedData = {
      _id: 'User|valid-user-2@manuscriptsapp.com',
      name: 'New Username',
      email: 'new-email@manuscriptsapp.com',
    }

    const user = await repository.update(userUpdatedData)
    expect(user).toEqual({
      _id: userUpdatedData._id,
      email: userUpdatedData.email,
      name: userUpdatedData.name,
    })

    await new Promise((resolve, reject) => {
      setTimeout(() => {
        repository
          .getById(userUpdatedData._id)
          .then((result) => {
            expect(result).toBeNull()
            resolve()
          })
          .catch((error) => reject(error))
      }, 2000)
    })
  })
})

describe('SQLRepository patch', () => {
  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Project)
    await seed({ users: true })
  })

  test('should fail if document _id does not match the id', () => {
    const repository = new UserRepository(db)
    const chance = new Chance()
    const id = chance.hash()
    return expect(
      repository.patch(id, { _id: chance.string(), name: chance.name() })
    ).rejects.toThrowError(ValidationError)
  })

  test('should fail if key does not exist', () => {
    const repository = new UserRepository(db)
    const chance = new Chance()
    const id = chance.hash()
    return expect(repository.patch(id, { name: chance.name() })).rejects.toThrowError(DatabaseError)
  })

  test('should fail if error occurred', () => {
    const repository = new UserRepository(db)
    const document = {
      name: 'long name used in test to generates error if name length is more than 100 char -  we need to fine new way to throw error from db internally- this just a tmp solution',
    }

    return expect(repository.patch(validUser2._id, document)).rejects.toThrowError(DatabaseError)
  })

  test('should patch user data successfully', async () => {
    const repository = new UserRepository(db)

    let user: any = await repository.patch('User|valid-user@manuscriptsapp.com', {
      name: 'Cody Rodriquez',
    })
    user = await repository.getById('User|valid-user@manuscriptsapp.com')

    expect(user.name).toBe('Cody Rodriquez')
  })

  test.skip('should delete the patched user after the expiry time pass', async () => {
    const repository = new UserRepository(db)

    const userUpdatedData = {
      _id: 'User|valid-user-2@manuscriptsapp.com',
      email: 'new-email@manuscriptsapp.com',
    }

    const user = await repository.patch(userUpdatedData._id, { email: userUpdatedData.email })
    expect(user.email).toEqual(userUpdatedData.email)
    await new Promise((resolve, reject) => {
      setTimeout(() => {
        repository
          .getById(userUpdatedData._id)
          .then((result) => {
            expect(result).toBeNull()
            resolve()
          })
          .catch((error) => reject(error))
      }, 2000)
    })
  })
})

describe('SQLRepository touch', () => {
  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Project)
    await seed({ users: true })
  })

  test('should fail if error occurred - key does not exist', () => {
    const repository = new UserRepository(db)
    const chance = new Chance()
    const id = chance.hash()
    return expect(repository.touch(id, 100)).rejects.toThrowError(DatabaseError)
  })

  test('should touch user successfully', async () => {
    const repository = new UserRepository(db)
    await repository.touch(validUser2._id, 100)
  })
})

describe('SQLRepository getById', () => {
  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Project)
    await seed({ users: true })
  })

  test('should fail if database.bucket not set', () => {
    const repository: any = new UserRepository(db)
    repository.database = {}
    const id = chance.hash()
    return expect(repository.getById(id)).rejects.toThrowError(NoBucketError)
  })

  test('should fail return null if key not exists', async () => {
    const repository = new UserRepository(db)
    const chance = new Chance()
    const id = chance.hash()
    const user = await repository.getById(id)

    expect(user).toBeNull()
  })

  test('should get user by id successfully', async () => {
    const repository = new UserRepository(db)
    const user: any = await repository.getById('User|valid-user@manuscriptsapp.com')

    expect(user._id).toBe('User|valid-user@manuscriptsapp.com')
    expect(user.email).toBe('valid-user@manuscriptsapp.com')
    expect(user.name).toBe('Valid System User')
  })
})

describe('SQLRepository getOne', () => {
  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Project)
    await seed({ users: true })
  })

  test('should fail if database.bucket not set', () => {
    const repository: any = new UserRepository(db)
    repository.database = {}
    const _id = chance.hash()
    return expect(repository.getOne({ _id })).rejects.toThrowError(NoBucketError)
  })

  test('should return undefined if no data exists', async () => {
    const repository = new UserRepository(db)
    const chance = new Chance()
    const user = await repository.getOne({ _id: chance.hash() })

    expect(user).toBeNull()
  })

  test('should return user', async () => {
    const repository = new UserRepository(db)

    const user = await repository.getOne({ _id: validUser2._id })
    expect(user).toMatchSnapshot()
  })
})

describe('SQLRepository count', () => {
  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Project)
    await seed({ users: true })
  })

  test('should fail return count as number there is keys', async () => {
    const repository = new UserRepository(db)

    const count = await repository.count({ _id: validUser2._id })
    expect(count).toBe(1)
  })

  test('should fail if database.bucket not set', () => {
    const repository: any = new UserRepository(db)
    repository.database = {}
    const _id = chance.hash()
    return expect(repository.count({ _id })).rejects.toThrowError(NoBucketError)
  })
})

describe('SQLRepository getAll', () => {
  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Project)
    await seed({ users: true })
  })

  test('should return empty set if no data exists', () => {
    const repository = new UserRepository(db)
    const chance = new Chance()
    const query = {
      email: chance.email(),
      name: chance.email(),
      _id: chance.hash(),
    }
    return expect(repository.getAll(query, null)).resolves.toEqual([])
  })

  test('should return user list', () => {
    const repository = new UserRepository(db)
    return expect(repository.getAll({ _id: validUser2._id }, null)).resolves.toMatchSnapshot()
  })

  test('should fail if database.bucket not set', () => {
    const repository: any = new UserRepository(db)
    repository.database = {}
    return expect(repository.getAll({ _id: validUser2._id }, null)).rejects.toThrowError(
      NoBucketError
    )
  })
})

describe('SQLRepository remove', () => {
  test('should remove key', async () => {
    const repository = new UserRepository(db)
    const query = {
      _id: validUser2._id,
    }
    await repository.remove(query)
    const user = await repository.getOne(query)
    expect(user).toBeNull()
  })

  test('should fail if database.bucket not set', () => {
    const repository: any = new UserRepository(db)
    repository.database = {}
    const query = { _id: validUser2._id }
    return expect(repository.remove(query)).rejects.toThrowError(NoBucketError)
  })
})

describe('SQLRepository buildModel', () => {
  test('should remove any property equal to undefined', () => {
    const repository = new UserRepository(db)
    const mappedModel = repository.buildModel({
      email: undefined,
      name: 'foo',
    } as any)

    expect(mappedModel.email).toBeUndefined()
    expect(mappedModel.name).toBeDefined()
  })

  test('should fail if database.bucket not set', () => {
    const repository: any = new UserRepository(db)
    repository.database = {}
    const query = { _id: validUser2._id }
    return expect(repository.remove(query)).rejects.toThrowError(NoBucketError)
  })
})
