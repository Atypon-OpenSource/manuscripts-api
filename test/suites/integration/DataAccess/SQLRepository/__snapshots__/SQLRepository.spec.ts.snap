// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SQLRepository getAll should return user list 1`] = `
[
  {
    "_id": "User|valid-user-2@manuscriptsapp.com",
    "connectUserID": undefined,
    "email": "valid-user-2@manuscriptsapp.com",
    "name": "Valid System User",
  },
]
`;

exports[`SQLRepository getOne should return user 1`] = `
{
  "_id": "User|valid-user-2@manuscriptsapp.com",
  "connectUserID": undefined,
  "email": "valid-user-2@manuscriptsapp.com",
  "name": "Valid System User",
}
`;
